dm_control/locomotion/walkers/assets/dog_v2/add_muscles.py

```diff
@@ -0,0 +1,435 @@
+    Returns:
+        closest_geom_body: The closest geometric body (as a Mujoco body object)
+        to the target point.
+  """
+  dist = 100000000000
+  closest_geom_body = 0
+  for bone, tree in kd_tree.items():
+    dist_new, i = tree.query(np.array([point]), k=1)
+    geom = mjcf.find("geom", bone)
+    body = geom.parent
+
+    if mtu in muscle_legs:
+      if body.name not in ["upper_leg_L",
+                           "upper_leg_R", "lower_leg_L", "lower_leg_R",
+                           "foot_L", "foot_R", "toe_L", "toe_R",
+                           "scapula_L", "scapula_R", "upper_arm_L",
+                           "upper_arm_R", "lower_arm_L", "lower_arm_R",
+                           "hand_L", "hand_R", "finger_L", "finger_R",
+                           "pelvis"]:
+        continue
+    else:
+      if body.name in ["upper_leg_L",
+                       "upper_leg_R", "lower_leg_L", "lower_leg_R",
+                       "foot_L", "foot_R", "toe_L", "toe_R",
+                       "lower_arm_L", "lower_arm_R",
+                       "hand_L", "hand_R", "finger_L", "finger_R"]:
+        continue
+
+    if dist_new < dist:
+      dist = dist_new
+      closest_geom_body = body
+
+  return closest_geom_body
+
+
+def calculate_transformation(element):
+  """
+    Calculates the transformation matrix from the root to the given element.
+
+    This function iterates through the parent elements of the given element and
+    calculates the transformation matrices from the root until reaching the
+    given element. It considers the position and orientation (quaternion)
+    attributes of each element to construct the transformation matrices.
+    The final transformation matrix represents the cumulative transformation
+    from the root to the given element.
+
+    Args:
+        element: The MuJoCo element for which the transformation matrix
+          is calculated.
+
+    Returns:
+        T: The transformation matrix from the root to the given element.
+  """
+  # Calculate all transformation matrices from root until this element
+  all_transformations = []
+
+  while type(element) == mjcf_module.element._ElementImpl:
+    if element.pos is not None:
+      pos = np.array(element.pos, dtype=float)
+    else:
+      pos = np.zeros(3)
+
+    if element.quat is not None:
+      rot = Rot.from_quat(element.quat).as_matrix()
+    else:
+      rot = Rot.identity().as_matrix()
+
+    all_transformations.append(
+      np.vstack(
+        (np.hstack((rot, pos.reshape([-1, 1]))), np.array([0, 0, 0, 1])))
+    )
+
+    element = element.parent
+
+  # Apply all transformations
+  T = np.eye(4)
+  for transformation in reversed(all_transformations):
+    T = np.matmul(T, transformation)
+
+  return T
+
+
+def add_muscles(model, scale_multiplier, muscle_dynamics,
+                asset_dir, lengthrange_from_joints):
+  physics = mjcf_module.Physics.from_mjcf_model(model)
+
+  mjcf = model
+
+  muscle_meshes_path = asset_dir + '/muscles/'
```

Comment:
nit: Generally, do join paths use the `os.path.join` as it's safer and handles different platforms better.